src/methodical/FnWithMeta.java

```diff
@@ -0,0 +1,51 @@
+package methodical;
+
+import clojure.lang.*;
+
+public class FnWithMeta extends AFunction {
```

Comment:
I'm really wary of adding a Java class here since adding ClojureScript support to Methodical is definitely something I want to pursue at some point in the future. Couldn't you just create a custom `deftype` that implements `clojure.lang.IFn` and `clojure.lang.IObject`?

Reply:
The problem is [RestFn](https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/RestFn.java) does more than that. It also implements a bunch of `doInvoke`s and helpers to make `apply` less expensive if the number of required arguments for the function is bigger than the number of supplied values. Rewriting that full class to a deftype would be daunting to say the least.

When Clojurescript support becomes desirable, you should be able to refactor the new functions (fn-with-meta, vary-fn-with-meta, unwrap) to defer to the plain `with-meta` for fn objects in ClojureScript, using reader conditionals.

Reply:
And `proxy` is very bad performance-wise.

Reply:
That class is 3k lines of code. I guess with macros it can be mostly generated. I can try to reverse engineer the full details of what RestFn is doing and write such a macro.